vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go

```diff
@@ -0,0 +1,162 @@
+	PodNamespaceLabel, PodNameLabel                  string
+	CtrNamespaceLabel, CtrPodNameLabel, CtrNameLabel string
+}
+
+// NewExternalClient returns a Source for an External Metrics Client.
+func NewExternalClient(c *rest.Config, clusterState *model.ClusterState, options ExternalClientOptions) PodMetricsLister {
+	extClient, err := external_metrics.NewForConfig(c)
+	if err != nil {
+		klog.Fatalf("Failed initializing external metrics client: %v", err)
+	}
+	return &externalMetricsClient{
+		externalClient: extClient,
+		options:        options,
+		clusterState:   clusterState,
+	}
+}
+
+type podContainerResourceMap map[model.PodID]map[string]k8sapiv1.ResourceList
+
+func (s *externalMetricsClient) addMetrics(list *externalmetricsv1beta1.ExternalMetricValueList, name k8sapiv1.ResourceName, resourceMap podContainerResourceMap, namespace, podName string) {
+	for _, val := range list.Items {
+		ctrName, hasCtrName := val.MetricLabels[s.options.CtrNameLabel]
+		if hasCtrName {
+			podId := model.PodID{
+				Namespace: namespace,
+				PodName:   podName,
+			}
+			if resourceMap[podId] == nil {
+				resourceMap[podId] = make(map[string]k8sapiv1.ResourceList)
+			}
+			if resourceMap[podId][ctrName] == nil {
+				resourceMap[podId][ctrName] = make(k8sapiv1.ResourceList)
+			}
+			resourceMap[podId][ctrName][name] = val.Value
+		}
+	}
+}
+
+func (s *externalMetricsClient) List(ctx context.Context, namespace string, state *model.ClusterState, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	result := v1beta1.PodMetricsList{}
+
+	for _, vpa := range state.Vpas {
+		if vpa.PodCount == 0 {
+			continue
+		}
+
+		workloadValues := make(podContainerResourceMap)
+		if namespace != "" && vpa.ID.Namespace != namespace {
+			continue
+		}
+
+		nsClient := s.externalClient.NamespacedMetrics(vpa.ID.Namespace)
+
+		var selectedTimestamp v1.Time
+		var selectedWindows time.Duration
+		for resourceName, metricName := range s.options.ResourceMetrics {
+			pods := s.clusterState.GetMatchingPods(vpa)
+			for _, pod := range pods {
+				podNameReq, err := labels.NewRequirement("pod", selection.Equals, []string{pod.PodName})
+				if err != nil {
+					return nil, err
+				}
+				selector := vpa.PodSelector.Add(*podNameReq)
+				m, err := nsClient.List(metricName, selector)
+				if err != nil {
+					return nil, err // Do we want to error or do we prefer to skip ?
+				}
+				if m == nil || len(m.Items) == 0 {
+					continue
+				}
+				klog.V(2).Infof("External Metrics Query for VPA %+v: resource %+v, metric %+v, %d items", vpa.ID, resourceName, metricName, len(m.Items))
+				s.addMetrics(m, resourceName, workloadValues, vpa.ID.Namespace, pod.PodName)
+
+				if !m.Items[0].Timestamp.Time.IsZero() {
+					selectedTimestamp = m.Items[0].Timestamp
+					if m.Items[0].WindowSeconds != nil {
+						selectedWindows = time.Duration(*m.Items[0].WindowSeconds) * time.Second
+					}
+				}
+			}
+
+		}
+
+		for podId, cmaps := range workloadValues {
```

Comment:
I don't understand why we copy metric values we got from `nsClient.List()` call to `workloadValues` map then copy from `workloadValues` to `result`. It seems to me that building `result.Items` immediately would be simpler.

And we can just copy `Timestamp` and `Window` for each sample and avoid the problem the comment "I am not sure that this is correct. Need to rework Timestamp and Window... which one should we use?" is talking about

Reply:
You're right.  There was a lot more here before w.r.t. other metircs it was putting out.  I'll merge it all together.